### DISCUSSION TITLE
Questions about electric field and weighting potential

### DISCUSSION THREAD

**Discussion by zql2021** [JuliaPhysics/SolidStateDetectors.jl#359] [Q&A] [answered]
Dear Sir, I have been using SolidStateDetectors for some simulations recently, but it seems that there are some issues with the visualization results of the electric field and weighting potential. 

1. when I calculate the weighting potential of contacts(id=2,3...18),all the result are zero
this is  my code and the visualization results.
`contact=sim.detector.contacts[12]
calculate_weighting_potential!(sim, contact.id, refinement_limits = [0.3, 0.3, 0.1])
plot(sim.weighting_potentials[12])`
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/6646ffa6-8f5a-494b-bee8-1811fafae34b)

2. And the electric field code and result
`simulate!(sim,device_array_type = CuArray,  refinement_limits = [0.3, 0.1, 0.05], max_distance_ratio = 4, max_n_iterations = 2000)
plot(sim.electric_field, full_det = true, φ = 0.0, cbar_lims=(0,100000),size = (700, 700))
plot_electric_fieldlines!(sim, full_det = true, φ = 0.0)`
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/ae5d96fc-fcb2-4dd0-945f-64510ddd5c39)
**Why are the electric field lines not uniform?**

3. Besides,I check the initial electric potential
`apply_initial_state!(sim, ElectricPotential)
plot(
    plot(sim.electric_potential), # initial electric potential (boundary conditions)
    plot(sim.point_types), # map of different point types: fixed point / inside or outside detector volume / depleted/undepleted
    plot(sim.q_eff_imp), # effective charge density distribution
    plot(sim.ϵ_r), # dielectric distribution
    layout = (2, 2), size = (1000, 500)
)
`and get the result below
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/a3a57923-a4eb-4009-aad1-4e50c890b58b)
**Why is the intial electric potential all zero?**

**This is my config file below**
[det_config.zip](https://github.com/JuliaPhysics/SolidStateDetectors.jl/files/12679269/det_config.zip)
Could you please take a look and provide some insights? Thank you very much and looking forward to your early reply.

**Comment by fhagemann** [comment, marked as answer]
I haven't run your simulations yet, but some first thoughts:

1. Does the contact with `id` 12 lie in the cross-section of `x = 0`. Maybe you are looking at a cross-section, where the weighting potential is actually (close to) zero.
2. When plotting the electric field lines, point charges are generated in the vicinity of all contacts that are not part of the `skip_contact` keyword. To get nice looking field lines, you might need to tune the plot using keyword arguments. You can look at the source code here: https://github.com/JuliaPhysics/SolidStateDetectors.jl/blob/main/src/PlotRecipes/ElectricField.jl#L90-L96
There is also some documentation here: https://juliaphysics.github.io/SolidStateDetectors.jl/stable/man/plotting/#Electric-Field-Plots
4. At the very beginning, the electric potential is chosen to be 0 everywhere as initial guess and then calculated using the SOR algorithm. Again, the electric potential might look like it's zero, because you might be plotting a cross-section where no contact has a fixed potential different from zero.

You can choose a different cross-section using e.g. `plot(.... , y = 0)`

**Reply by zql2021** [reply]
Thank you for your timely reply!
Question1：The contact with id 12 doesn't lie in the cross-section of x = 0.
The picture of the detector is as follows. The blue one is the guard ring(contact id =1and id=9 in two sides respectively),the yellow ones are the electrodes(id=2,3,4...8，id=9,10,11...18)
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/4610c966-0168-4022-927e-07a27b1b96f1)
I have applied voltage to the guard ring and 8 electrodes on one side.

Question2&3：Thank you for your helpful  advice, I will  try them.

**Reply by fhagemann** [reply]
1. If the contact is not in the cross section, you might maybe only expect very weak weighting potentials. You could try adjusting the limits of the colobar (clims = (0,0.1)), so see whether there is something or plot equipotential lines (contours_equal_potential = true, levels = 0:0.005:1, linecolor = :white)

**Reply by zql2021** [reply]
Dear Sir, I have identified where the issue was. The contacts were along the y-axis, while the default view of the plot was the y-z plane. I just adjusted the electrode orientation along the x-axis, and the visualization is now correct.

1. So, my other question is how to adjust the visualization view if I want to plot the x-z plane?
2. When I import Geant4 simulation results into SSD and  visualize them, I get four graphs once. If I only want to work with one of the first three graphs, do you have any suggestions? 
`plot(mc_events_clustered)`
<img width="438" alt="image" src="https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/4ccfbed5-60d4-4e41-9a16-b1a791df9783">

**Reply by fhagemann** [reply]
1. In the `plot` command, you can select a `x`-`z` plane by calling it with the `y` value at which you want the cross-section to be plotted, e.g. `plot(...., y = 0)` to plot it in the `y = 0`-plane.
2. The plot you are seeing is a custom one made for `DetectorHitEvents` which is defined in RadiationDetectorSignals.jl (https://github.com/JuliaPhysics/RadiationDetectorSignals.jl/blob/main/src/detector_hits.jl) and imported into SolidStateDetectors.jl.
The corresponding plot recipe can be found here: https://github.com/JuliaPhysics/RadiationDetectorSignals.jl/blob/main/src/plots_recipes.jl
One quick fix would be to overwrite the plot recipe for `DetectorHitEvents` to only show the top left plot, e.g.
```julia
using RecipesBase
using SolidStateDetectors: DetectorHitEvents, flatview

@recipe function f(events::DetectorHitEvents)

    edep_flat = collect(flatview(events.edep))
    edep_unit = unit(eltype(edep_flat))
    edep_unitless = Array(ustrip(edep_flat))

    unitformat --> :square

    pos_flat = flatview(collect(flatview(events.pos)))
    pos_x = pos_flat[1,:]
    pos_y = pos_flat[2,:]
    pos_z = pos_flat[3,:]

    @series begin
        seriestype := :histogram2d
        weights := edep_unitless
        aspect_ratio := 1.0

        bins --> 500

        #title := "Detector Hits, XY"
        xguide := "x"
        yguide := "y"
        

        (pos_x, pos_y)
    end
end
```
Then, `plot(mc_events_clustered)` should give you the desired plot.

**Reply by zqiuli** [reply]
I understand , and  I greatly appreciate your detailed response!
